trieste/acquisition/function.py

```diff
@@ -298,3 +298,96 @@ def probability_of_feasibility(
     return distr.cdf(tf.cast(threshold, at.dtype))
+
+
+class ExpectedConstrainedImprovement(AcquisitionFunctionBuilder):
+    """
+    Builder for the _expected constrained improvement_ acquisition function defined in
+    Gardner, 2014. The acquisition function computes the expected improvement from the best
+    feasible point, where feasible points are those that (probably) satisfy some constraint.
+
+    See the following for details:
+
+    ::
+
+    @inproceedings{gardner14,
+        title={Bayesian Optimization with Inequality Constraints},
+        author={Jacob Gardner and Matt Kusner and Zhixiang and Kilian Weinberger and John Cunningham},
+        booktitle={Proceedings of the 31st International Conference on Machine Learning},
+        year={2014},
+        volume={32},
+        number={2},
+        series={Proceedings of Machine Learning Research},
+        month={22--24 Jun},
+        publisher={PMLR},
+        url={http://proceedings.mlr.press/v32/gardner14.html},
+    }
+
+    """
+
+    _PENALIZATION_FACTOR = 1.1
+
+    def __init__(
+        self,
+        objective_tag: str,
+        constraint_builder: AcquisitionFunctionBuilder,
+        min_feasibility_probability: Union[float, tf.Tensor] = 0.5,
+    ):
+        """
+        :param objective_tag: The tag for the objective data and model.
+        :param constraint_builder: The builder for the constraint function.
+        :param min_feasibility_probability: The minimum probability of feasibility for a
+            "best point" to be considered feasible.
+        :raise ValueError (or InvalidArgumentError): If ``min_feasibility_probability`` is not a
+            scalar in the unit interval :math:`[0, 1]`.
+        """
+        tf.debugging.assert_scalar(min_feasibility_probability)
+
+        if not 0 <= min_feasibility_probability <= 1:
+            raise ValueError(
+                f"Minimum feasibility probability must be between 0 and 1 inclusive,"
+                f" got {min_feasibility_probability}"
+            )
+
+        self._objective_tag = objective_tag
+        self._constraint_builder = constraint_builder
+        self._min_feasibility_probability = min_feasibility_probability
+
+    def prepare_acquisition_function(
+        self, datasets: Mapping[str, Dataset], models: Mapping[str, ModelInterface]
+    ) -> AcquisitionFunction:
+        """
+        :param datasets: The data from the observer.
+        :param models: The models over each dataset in ``datasets``.
+        :return: The expected constrained improvement acquisition function.
+        :raise KeyError: If `objective_tag` is not found in ``datasets`` and ``models``.
+        :raise ValueError: If the objective data is empty, or no feasible points are found in
+            ``datasets``.
+        """
+        objective_model = models[self._objective_tag]
+        objective_dataset = datasets[self._objective_tag]
+
+        if len(objective_dataset.query_points) == 0:
```

Comment:
can do `len(dataset) == 0`

Reply:
actually we need here `query points` to compute  `mean`

Reply:
len query_points is the same as len dataset